accepted/wasm-gc.md

```diff
@@ -0,0 +1,1319 @@
+    /// Write barrier for array elements: `array[index] = value`.
+    ///
+    /// This is only called when `GcMetadata::needs_write_barriers == true`,
+    /// otherwise Wasmtime will just do a raw write inline.
+    ///
+    /// Parameters:
+    ///
+    /// * `heap`: The heap that the array is within.
+    ///
+    /// * `recever_layout`: The layout of the array.
+    ///
+    /// * `array`: A pointer to the array.
+    ///
+    /// * `index`: The index being written to (note that this is *not* the
+    ///   element's offset, that can be obtained via
+    ///   `array_layout.header_size + index * sizeof(element)`).
+    ///
+    /// * `value`: The value being written into the element.
+    unsafe fn write_array(
+        &self,
+        heap: GcHeap,
+        array_layout: &wasmtime_environ::GcObjectLayout,
+        array: NonNull<u8>,
+        element: u32,
+        value: ValRaw,
+    );
+
+    /// Read barrier for array elements: `array[index]`.
+    ///
+    /// This is only called when `GcMetadata::needs_read_barriers == true`,
+    /// otherwise Wasmtime will just do a raw read inline.
+    ///
+    /// Parameters:
+    ///
+    /// * `heap`: The heap that the array is within.
+    ///
+    /// * `recever_layout`: The layout of the array.
+    ///
+    /// * `array`: A pointer to the array.
+    ///
+    /// * `index`: The index being read from. (Note that `index` is *not*
+    ///   the offset of the index, but an index of the `i`th index of the
+    ///   object. The offset can be obtained via
+    ///   `array_layout.indexs[index].0`.)
+    ///
+    /// Result:
+    ///
+    /// * The value read from the index.
+    fn read_array(
+        &self,
+        heap: GcHeap,
+        array_layout: &wasmtime_environ::GcArrayLayout,
+        array: NonNull<u8>,
+        index: u32,
+    ) -> ValRaw;
+}
+```
+
+It isn't 100% clear to me that we want the whole
+`GcMetadata::need_{read,write}_barriers` mechanism. It is a space-time trade
+off: bloating `StructRef` and `ArrayRef` with a cached copy of those `bool`s to
+avoid a virtual call. Inside `wasmtime::StructRef::get` would be something like
+this:
+
+```rust
+if !self.cached_need_read_barriers {
+    do_raw_read_inline();
+} else {
+    do_virtual_call_barrier();
+}
+```
+
+This is making the assumption that the well-predicted branch with an inline read
+is faster than a well-predicted virtual call with an out-of-line read. The
+scheme also imposes a little bit of extra complexity that isn't strictly
+functionally necessary. So I'm on the fence about it, we could easily remove it
+and always do the virtual call. Or we could start with always doing the virtual
+call and add this scheme later if performance becomes a concern.
+
+## GC Implementations
+
+We will have three GC implementations:
+
+1. The null GC, which allocates until it runs out of heap space, at which
+   point it triggers a trap in the guest.
+
+2. A general-purpose collector, built on top of an off-the-shelf GC library.
```

Comment:
Out of curiosity, is the "GC" part of the off-the-shelf library necessary? It seems like we're doing all the hard work of rooting and maintaining layouts, so building an implementation on top of `malloc` wouldn't really be all that much worse than any other GC library at that point (like boehm or something like that I'd presume).

This definitely makes sense to have, but it seems to be like it could be basically something backed by `SystemAlloc` in Rust.

Reply:
The hard part is not rooting and object layouts, but implementing a nursery, compaction, background sweeping, parallel tracing, incremental GC, and everything else that a general-purpose GC is going to eventually want.

Reply:
(still have some thoughts so un-resolving this temporarily)

For a first-pass implementation do we need that though? Of the benefits you listed:

* Nursery / compaction - these seem like things that `malloc` is a "proxy" for, although I realize that `malloc` is not perfect for these of course. 
* Background sweeping - we're not going to want this any time in the near future I think? That'd be reliant on something like the `shared` attribute in core wasm?
* Parallel tracing - does a GC library do this for us? Wouldn't we do all the tracing ourselves since we're the ones understanding the heap layout of all GC objects? I haven't worked with any off-the-shelf GC libraries so I'm likely missing something here.

I can understand how perhaps, one day, we want a super-fancy all-the-bells-and-whistles GC. That to me feels on the order of numbers-of-years out though so it seems like "just call `malloc`" with a simple mark/sweep could be sufficient for quite some time.

Reply:
> Nursery / compaction - these seem like things that malloc is a "proxy" for, although I realize that malloc is not perfect for these of course.

Do you mean proxy in that the `malloc` implementation should handle these things for us? I don't think it can or will. I'm not totally following you here.

> Background sweeping - we're not going to want this any time in the near future I think? That'd be reliant on something like the shared attribute in core wasm?

Because wasm-gc doesn't have finalizers, it isn't semantically visible to the wasm program where/when sweeping happens. Therefore, it doesn't rely on anything like a `shared` attribute, and is purely a runtime optimization. And it is a very nice optimization, since sweeping is O(heap size) not O(live objects), at least in simple mark-sweep.

> Parallel tracing - does a GC library do this for us? Wouldn't we do all the tracing ourselves since we're the ones understanding the heap layout of all GC objects? I haven't worked with any off-the-shelf GC libraries so I'm likely missing something here.

It depends on the details of the library, I suppose. I haven't used any of these libraries in anger myself either.

But before digging in too much further: we are *not* the ones who understand object layout. It is the collector that knows object layout, and us who have to indirectly access fields. And object layout *has* to be the collector's responsibility since different collectors have different requirements for things like header words or reference counts or any other kind of metadata it requires.

> I can understand how perhaps, one day, we want a super-fancy all-the-bells-and-whistles GC. That to me feels on the order of numbers-of-years out though so it seems like "just call malloc" with a simple mark/sweep could be sufficient for quite some time.

Yeah, perhaps if we don't hear from any of the non-FaaS Wasmtime embedders (who are all going to want to use the semi-space copying collector) we can ignore the general-purpose GC usecase for now and only have the null collector and the semi-space copying collector.

But if we still want to have a general-purpose GC, then I'm not sure what you are proposing. Just doing `malloc`/`free` and simple mark-sweep ourselves is going to be a very suboptimal collector, probably worse than the semi-space copying collector for long-running Wasm programs. So I'm not sure it would actually gain us anything.

Reply:
Hm before I answer some other things one thing you said confuses me:

> And object layout has to be the collector's responsibility since different collectors have different requirements for things like header words or reference counts or any other kind of metadata it requires.

I don't understand this because doesn't compiled code need to access the object's layout? Doesn't this mean that, no matter what, the in-memory layout of an object is intimately know by the runtime to something other than a collector as a black box? For example the `GcCompiler` trait would have a very close relationship to the collector, so if we picked an off-the-shelf library doing fancy things here we'd reverse-engineer it and codify it in a corresponding `GcCompiler` implementation. 

I think overall what I'm feeling is that we should start with something backed by mark-and-sweep plus `malloc` knowing that it's not going to be that performant. We can have a specialization for the pooling allocator and FaaS-style runtimes of course, but I'm hesitant to bring in a large GC library dependency. GC is notoriously hard to get right and I'm not sure that any of us are so familiar with any particular GC library that we'd be willing to stake our security on the correctness of it and our understanding of its internals.

Reply:
> I don't understand this because doesn't compiled code need to access the object's layout? Doesn't this mean that, no matter what, the in-memory layout of an object is intimately know by the runtime to something other than a collector as a black box? For example the GcCompiler trait would have a very close relationship to the collector, so if we picked an off-the-shelf library doing fancy things here we'd reverse-engineer it and codify it in a corresponding GcCompiler implementation.

Yes, object layout would be completely encapsulated within `GcCompiler` implementations. The `GcCompiler` implementation wouldn't really "reverse engineer" the object layout, but would be considered "part of" the collector and would have insider knowledge/collaboration that the rest of the compiler/runtime wouldn't have (or would only get indirectly via `GcObjectLayout`) and which would potentially differ across different collector implementations.

> I think overall what I'm feeling is that we should start with something backed by mark-and-sweep plus malloc knowing that it's not going to be that performant. We can have a specialization for the pooling allocator and FaaS-style runtimes of course, but I'm hesitant to bring in a large GC library dependency. GC is notoriously hard to get right and I'm not sure that any of us are so familiar with any particular GC library that we'd be willing to stake our security on the correctness of it and our understanding of its internals.

I agree with you about the difficulty of building a correct and performant GC.

What I'm saying though is that if we don't want to pull in an external library for the general-purpose GC use case, then we shouldn't attempt to build our own. Instead we should just use the semi-space copying collector that we need for pooling allocator integration for the general-purpose GC use case. It won't be as good as what someone could build with MMTK, but it will be better than naive mark-and-sweep backed by malloc/free. And this approach allows us to build just two GCs (null, semi-space copying) instead of three (null, semi-space copying, mark-and-sweep).

Reply:
I've updated the RFC to remove MMTK and the general-purpose collector. Now there is just the null GC and the semi-space copying collector.

Reply:
Ok yeah that all sounds reasonable to me 👍